packages/backend-common/src/database/DatabaseManager.ts

```diff
@@ -0,0 +1,160 @@
+import {
+  createDatabaseClient,
+  ensureDatabaseExists,
+  createNameOverride,
+  normalizeConnection,
+} from './connection';
+import { PluginDatabaseManager } from './types';
+
+function pluginPath(pluginId: string): string {
+  return `plugin.${pluginId}`;
+}
+
+export class DatabaseManager {
+  /**
+   * Creates a DatabaseManager from `backend.database` config.
+   *
+   * The database manager allows the user to set connection and client settings on a per pluginId
+   * basis by defining a database config block under `plugin.<pluginId>` in addition to top level
+   * defaults. Optionally, a user may set `prefix` which is used to prefix generated database
+   * names if config is not provided.
+   *
+   * @param config The loaded application configuration.
+   */
+  static fromConfig(config: Config): DatabaseManager {
+    const databaseConfig = config.getConfig('backend.database');
+
+    return new DatabaseManager(
+      databaseConfig,
+      databaseConfig.getOptionalString('prefix'),
+    );
+  }
+
+  private constructor(
+    private readonly config: Config,
+    private readonly prefix: string = 'backstage_plugin_',
+  ) {}
+
+  /**
+   * Generates a PluginDatabaseManager for consumption by plugins.
+   *
+   * @param pluginId The plugin that the database manager should be created for. Plugin names should be unique
+   * as they are used to look up database config overrides under `backend.database.plugin`.
+   */
+  forPlugin(pluginId: string): PluginDatabaseManager {
+    const _this = this;
+
+    return {
+      getClient(): Promise<Knex> {
+        return _this.getDatabase(pluginId);
+      },
+    };
+  }
+
+  /**
+   * Provides the canonical database name for a given pluginId.
+   *
+   * This method provides the effective database name which is determined using global
+   * and plugin specific database config. If no explicit database name is configured,
+   * this method will provide a generated name which is the pluginId prefixed with
+   * 'backstage_plugin_'.
+   *
+   * @param pluginId Lookup the database name for given plugin
+   */
+  private getDatabaseName(pluginId: string): string {
+    const pluginConfig: Config = this.getConfigForPlugin(pluginId);
+
+    // determine root sqlite config to pass through as this is a special case
+    const rootConnection = this.config.get('connection');
+    const rootSqliteName =
+      typeof rootConnection === 'string'
+        ? rootConnection
+        : this.config.getOptionalString('connection.filename') ?? ':memory:';
+
+    const isSqlite = this.config.getString('client') === 'sqlite3';
+    return (
+      // attempt to lookup pg and mysql database name
+      pluginConfig.getOptionalString('connection.database') ??
+      // attempt to lookup sqlite3 database file name
+      pluginConfig.getOptionalString('connection.filename') ??
+      // if root is sqlite - attempt to use top level connection, fallback to :memory:
+      (isSqlite ? rootSqliteName : null) ??
```

Comment:
Addressed, I've split up the plugin block so that there are clear logical branches to follow.

https://github.com/backstage/backstage/pull/5700/commits/134f98d0fa12cb7e5ab6ca06ebe81b1128e19b44